icicle_v3/backend/cpu/src/hash/poseidon/cpu_poseidon.h

```diff
@@ -0,0 +1,203 @@
+#include <cassert>
+#include "icicle/fields/field.h"
+#include "hash.h"                   // DEBUG - use common hash.h and not from the local dir.
+#include "icicle/fields/field_config.h"
+
+// typedef uint32_t limb_t;            // DEBUG. To remove after Hadar submit hash.h.
+
+namespace icicle {
+ 
+/**
+ * @brief Poseidon hash constant struct.
+ */
+template <typename S>
+struct PoseidonConstants {
+  const S* m_rounds_constants;     ///< Rounds constants of this Poseidon hash.
+  const S* m_mds_matrix;           ///< Full rounds MDS matrix.
+  const S* m_pre_matrix;           ///< Pre-matrix (used at the last round of the upper full rounds).
+  const S* m_sparse_round_matrix;  ///< Partial rounds MDS matrix.
+  PoseidonConstants(const S* m_rounds_constants, const S* m_mds_matrix, const S* m_pre_matrix, const S* m_sparse_round_matrix) :
+    m_rounds_constants(m_rounds_constants), m_mds_matrix(m_mds_matrix),
+    m_pre_matrix(m_pre_matrix), m_sparse_round_matrix(m_sparse_round_matrix) {}
+};
+
+/**
+ * @brief Poseidon hash class.
+ */
+// template <typename S = scalar_t>
+template <typename S>
+class Poseidon : public Hash {
+  public:
+    Poseidon (
+      const unsigned int  arity,
+      const unsigned int  alpha,
+      const unsigned int  nof_partial_rounds,
+      const unsigned int  nof_upper_full_rounds,
+      const unsigned int  nof_end_full_rounds,
+      const S*            rounds_constants,
```

Comment:
if you provide 4 arrays, why do you need the struct PoseidonConstants?

Reply:
You wanted to put all matrices and round consts in a structure